modules/imgcodecs/src/grfmt_spng.cpp

```diff
@@ -0,0 +1,486 @@
+
+                if (ihdr.bit_depth <= 8 || ihdr.bit_depth == 16) {
+                    int num_trans = 0;
+                    switch (ihdr.color_type) {
+                        case SPNG_COLOR_TYPE_TRUECOLOR:
+                        case SPNG_COLOR_TYPE_INDEXED:
+                            struct spng_trns trns;
+                            num_trans = !spng_get_trns(ctx, &trns);
+                            if (num_trans > 0)
+                                m_type = CV_8UC4;
+                            else
+                                m_type = CV_8UC3;
+                            break;
+                        case SPNG_COLOR_TYPE_GRAYSCALE_ALPHA:
+                        case SPNG_COLOR_TYPE_TRUECOLOR_ALPHA:
+                            m_type = CV_8UC4;
+                            break;
+                        default:
+                            m_type = CV_8UC1;
+                    }
+                    if (ihdr.bit_depth == 16)
+                        m_type = CV_MAKETYPE(CV_16U, CV_MAT_CN(m_type));
+                    result = true;
+                }
+            }
+        }
+
+        if( !result )
+            close();
+
+        return result;
+    }
+
+
+    bool  SPngDecoder::readData( Mat& img )
+    {
+        volatile bool result = false;
+        bool color = img.channels() > 1;
+
+        struct spng_ctx *png_ptr = (struct spng_ctx *)m_ctx;
+
+        if( m_ctx && m_width && m_height )
+        {
+            int fmt = SPNG_FMT_PNG;
+
+            struct spng_trns trns;
+            int have_trns = !spng_get_trns((struct spng_ctx*) m_ctx, &trns);
+
+            int decode_flags = 0;
+            if(have_trns || img.channels() == 4)
+            {
+                if( m_color_type == SPNG_COLOR_TYPE_TRUECOLOR ||
+                    m_color_type == SPNG_COLOR_TYPE_INDEXED ||
+                    m_color_type == SPNG_COLOR_TYPE_TRUECOLOR_ALPHA)
+                    fmt = m_bit_depth == 16 ? SPNG_FMT_RGBA16 : SPNG_FMT_RGBA8;
+                else if( m_color_type == SPNG_COLOR_TYPE_GRAYSCALE )
+                    fmt = m_bit_depth == 16 ? SPNG_FMT_GA16 : SPNG_FMT_GA8;
+                else
+                    fmt = SPNG_FMT_RGBA8;
+
+                decode_flags = SPNG_DECODE_TRNS;
+            }
+
+            if(img.channels() < 4)
+            {
+                if( m_color_type == SPNG_COLOR_TYPE_INDEXED ||
+                    m_color_type == SPNG_COLOR_TYPE_TRUECOLOR ||
+                    m_color_type == SPNG_COLOR_TYPE_TRUECOLOR_ALPHA ||
+                    (m_color_type == SPNG_COLOR_TYPE_GRAYSCALE && img.channels() == 3))
+                    fmt = SPNG_FMT_RGB8;
+                else if( m_color_type == SPNG_COLOR_TYPE_GRAYSCALE &&
+                         m_bit_depth <= 8 )
+                    fmt = SPNG_FMT_G8;
+            }
+
+            size_t image_width, image_size = 0;
+            int ret = spng_decoded_image_size(png_ptr, fmt, &image_size);
+
+            if(!ret) {
```

Comment:
the code is correct, but still quite difficult to read. Could you please use explicit `if (ret == SPNG_OK) { ... }` to make the code easier to read?

Reply:
Done